File: src/main/java/com/indiepost/controller/admin/AdminContributorController.java
package com.indiepost.controller.admin;

import com.indiepost.dto.ContributorDto;
import com.indiepost.dto.DeleteResponse;
import com.indiepost.enums.Types;
import com.indiepost.service.ContributorService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/admin/contributors")
public class AdminContributorController {

    private final ContributorService contributorService;

    @Autowired
    public AdminContributorController(ContributorService contributorService) {
        this.contributorService = contributorService;
    }

    @GetMapping
    public Page<ContributorDto> get(@RequestParam String type, Pageable pageable) {
        if (type != null) {
            return contributorService.find(Types.ContributorType.valueOf(type), pageable);
        }
        return contributorService.find(pageable);
    }

    @GetMapping("/{id}")
    public ContributorDto get(@PathVariable Long id) {
        return contributorService.findOne(id);
    }

    @PostMapping
    public ContributorDto create(@RequestBody ContributorDto dto) {
        return contributorService.save(dto);
    }

    @PutMapping("/{id}")
    public ContributorDto update(@PathVariable Long id, @RequestBody ContributorDto dto) {
        if (dto.getId() == null || !dto.getId().equals(id)) {
            dto.setId(id);
        }
        return contributorService.save(dto);
    }

    @DeleteMapping("/{id}")
    public DeleteResponse delete(@PathVariable Long id) {
        Long deletedId = contributorService.deleteById(id);
        return new DeleteResponse(deletedId);
    }

}

File: src/main/kotlin/com/indiepost/controller/admin/AdminContributorController.kt
package com.indiepost.controller.admin

import com.indiepost.dto.ContributorDto
import com.indiepost.dto.DeleteResponse
import com.indiepost.enums.Types
import com.indiepost.service.ContributorService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.web.bind.annotation.*
import javax.inject.Inject
import javax.validation.ValidationException

@RestController
@RequestMapping("/admin/contributors")
class AdminContributorController @Inject
constructor(private val contributorService: ContributorService) {

    @GetMapping
    fun getList(@RequestParam type: String?, pageable: Pageable): Page<ContributorDto> {
        return if (type != null) {
            contributorService.find(Types.ContributorType.valueOf(type), pageable)
        } else contributorService.find(pageable)
    }

    @GetMapping("/{id}")
    fun getContributor(@PathVariable id: Long): ContributorDto? {
        return contributorService.findOne(id)
    }

    @PostMapping
    fun create(@RequestBody dto: ContributorDto): ContributorDto {
        return contributorService.save(dto)
    }

    @PutMapping("/{id}")
    fun update(@PathVariable id: Long, @RequestBody dto: ContributorDto): ContributorDto {
        if (dto.id != id) {
            throw ValidationException("Path variable and dto.id does not match: path($id) != dto(${dto.id})")
        }
        return contributorService.save(dto)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Long): DeleteResponse {
        val deletedId = contributorService.deleteById(id)
        return DeleteResponse(deletedId!!)
    }

}
